### PR TITLE
Disallow public state variables overriding pure functions.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
I think this is totally a bugfix we want to have. It is interesting that we already have several tests that now add the new error! The problem is that in order to fix these tests, we make the base contract functions `view` instead of `pure`, but then we get many warnings: state mutability can be restricted to `view`.

Should we not issue this warning for a `virtual` function?

**Comment by chriseth** [comment]
Added a second commit: Functions reading constant state variables are considered `pure` - so I think constant public state variables should also be pure.

**Review by ekpyron — COMMENTED** [review]
Yep, we should definitely have this - and the change looks fine, so "soft"-approval modulo the question of the tests.

**Comment by ekpyron** [comment]
> I think this is totally a bugfix we want to have. It is interesting that we already have several tests that now add the new error! The problem is that in order to fix these tests, we make the base contract functions `view` instead of `pure`, but then we get many warnings: state mutability can be restricted to `view`.
> 
> Should we not issue this warning for a `virtual` function?

Yeah, not issuing the warning for ``virtual`` functions makes sense. We already don't issue it for functions with empty implementation - basically because we didn't *have* ``virtual`` at the time - we could think about removing that special case and instead not warn for anything ``virtual``. (A non-virtual function with empty implementation makes little sense anyways, doesn't it?)

**Comment by chriseth** [comment]
Pull request is here: https://github.com/ethereum/solidity/pull/9418

File: libsolidity/analysis/OverrideChecker.cpp [hunk 1]
```diff
@@ -584,35 +584,35 @@ void OverrideChecker::checkOverride(OverrideProxy const& _overriding, OverridePr
...
 
-			if (_overriding.unimplemented() && !_super.unimplemented())
-				overrideError(
-					_overriding,
-					_super,
-					4593_error,
-					"Overriding an implemented function with an unimplemented function is not allowed."
-				);
-		}
+		if (_overriding.isFunction() && _overriding.unimplemented() && !_super.unimplemented())
```
Comment (Marenz) [inline libsolidity/analysis/OverrideChecker.cpp]: This might not be needed as the proxy always returns `false` for state variables

**Review by Marenz — CHANGES_REQUESTED** [review]

**Comment by chriseth** [comment]
Ready for review.

**Comment by Marenz** [comment]
Tests are failing though?

File: libsolidity/analysis/OverrideChecker.cpp [hunk 2]
```diff
@@ -584,35 +584,35 @@ void OverrideChecker::checkOverride(OverrideProxy const& _overriding, OverridePr
 				"Overridden " + _overriding.astNodeName() + " is here:"
 			);
 
-		// This is only relevant for a function overriding a function.
-		if (_overriding.isFunction())
-		{
-			// Stricter mutability is always okay except when super is Payable
-			if ((
+		// Stricter mutability is always okay except when super is Payable
```
Comment (Marenz) [inline libsolidity/analysis/OverrideChecker.cpp:587]: I was thinking that you don't actually need to change this whole block at all other than adding the `|| isVariable` part.
Is there an advantage of having `if (_overriding.unimplemented() && !_super.unimplemented())` outside the if-block?

The `if` would be slightly easier to read if we just do a small change

File: docs/contracts/inheritance.rst [hunk 3]
```diff
@@ -310,7 +310,8 @@ of the variable:
 
     contract A
     {
-        function f() external pure virtual returns(uint) { return 5; }
+        uint private num = 5;
+        function f() external view virtual returns(uint) { return num; }
```
Comment (chriseth) [inline docs/contracts/inheritance.rst]: This needs the PR that virtual suppresses the warning.

**Review by Marenz — APPROVED** [review]

**Pull request merged**